Added helm values, templates, and readme

miles-w-3 opened the pull request:
Closes #960 by adding structure to the helm `values.yaml`. Added functionality to simplify linking to an external postgres module using a helm template. This creates a pattern that can be used to simplify other processes, like using a PVC for external storage.

I would be happy to iterate on the helm values format or add additional functionality if there is interest. I also think that this should be merged after PR #954, and I can rebase the branch off of those changes once approved. This allows us to fully leverage the unified functionality of `$.Release.Namespace`

miles-w-3 committed:
Added helm values, templates, and readme

shanemcd commented:
![](https://i.kym-cdn.com/photos/images/newsfeed/000/117/814/are-you-wizard.jpg)

shanemcd commented on .helm/starter/templates/awx-deploy.yaml:
```diff
@@ -0,0 +1,24 @@
+  postgres_configuration_secret: {{ include "postgres.secretName" $ }}
+  {{- end }}
+{{- end }}
+{{- end }}
```
Can you configure your editor to insert trailing newlines and fix the 4 files in this PR that have this pesky warning?

miles-w-3 replied:
Sure, I'll get on that later today

shanemcd commented:
Running CI checks against https://github.com/ansible/awx-operator/pull/954 now and enabled auto-merge on that PR.

miles-w-3 commented:
Cool, I will rebase against those changes and make sure all of my changes are compatible once that's been merged in.

miles-w-3 commented:
@shanemcd I believe I've fixed the newlines, going to take a look at the #954 pipeline to see why it's failing

miles-w-3 commented:
Added updates to the starter readme as well as the main readme. Rebased onto @bewing's branch that includes the update -n awx to the workflwo

shanemcd commented on README.md:
```diff
@@ -253,6 +253,8 @@ For an example using the Nginx Controller in Minukube, don't miss our [demo vide
 
 For those that wish to use [Helm](https://helm.sh/) to install the awx-operator to an existing K8s cluster:
 
+The helm chart is generated from the `helm-chart` Makefile section using the starter files in `.helm/starter`. Consult [the documentation](.helm/starter/README.md) on how to customize the AWX resource with your own values.
```
👍

shanemcd commented:
The test output is hard to read, but search the raw output for "molecule test -s kind" and you'll see yamllint is failing. I can't even tell what file it's complaining about. Disabling colored output might help. If not, maybe try running it locally. 🤷

miles-w-3 commented:
I'll take a look, also if #954's workflow works this time I can base my branch off of it directly to make sure it's as accurate as possible

miles-w-3 commented:
Ok in my branch I added `.helm/starter` to the `.yamllint` ignore config. I think this makes sense because there is already a helm section set up. To compensate, I added some new validation steps to the helm validation section:
1. Run `helm lint` against the config - this provides some limited checks, especially to syntax and will probably have more descriptive error messages
2. Run kubeval: This is a much more rigorous set of tests designed to make sure that kubernetes manifests follow the given schema - Secret, ClusterRole, etc.  To use with helm, i create an output of all of the helm templates using `helm template`. This has the added benefit of debugging the schemas. Kubeval also has special integration with the templated file, it will use the comments to trace the error back to the template that is created. [Here](https://www.kubeval.com/#helm) is the link to the kubeval site 

 I hope this is enough testing and makes up for removing the other molecule tests - I think it makes sense to break them up fully because they are different uses of yaml. Let me know if you'd like me to add more tests

janorn commented:
Perhaps this could be useful to validate the chart?

https://github.com/sbaudoin/yamllint/issues/16#issuecomment-1018604978

miles-w-3 commented:
> Perhaps this could be useful to validate the chart?
> 
> [sbaudoin/yamllint#16 (comment)](https://github.com/sbaudoin/yamllint/issues/16#issuecomment-1018604978)

Yeah, we could also run yamllint against the template file, but I would argue that kubeval accomplishes and extends that functionality. The lines I have added look very similar to what you linked:
```
helm template -n awx awx-operator > tmp/test.yaml
kubeval --strict --force-color --ignore-missing-schemas tmp/test.yaml
```

miles-w-3 commented:
Cool, tests passed. @shanemcd would you be willing to merge the create namespace PR branch?

shanemcd commented:
@miles-w-3 Apologies for the delay, I had to prioritize some other stuff for a while. I've merged that PR. Once you fix the conflicts here we'll get this one in. Thanks for all your work.

miles-w-3 commented:
No problem, rebasing off of new devel now and will update the PR

miles-w-3 commented:
@shanemcd rebased off of devel, [passed tests on my fork](https://github.com/miles-w-3/awx-operator/runs/7292829395?check_suite_focus=true) so should be ready to go. 
Out of curiosity, how often do you do releases, and how often do you push the packaged helm chart to the image host?
Also, if you have more features you'd like to see in the helm chart I would be happy to help out

shanemcd commented:
> Out of curiosity, how often do you do releases, and how often do you push the packaged helm chart to the image host?

Roughly every 3 weeks but we dont have a schedule set in stone as sometimes we have to prioritize working on stuff for customers. A release went out just now but we can do another one ad-hoc for this if it would help you.

The helm chart gets automatically published during the release process:

https://github.com/ansible/awx-operator/blob/7697825944e7bc7f7c3b67f1e5b87f70dd42c8cf/.github/workflows/promote.yaml#L34-L40

miles-w-3 commented:
That would be great if you wouldn't mind, these changes would definitely be useful for me to implement/document for my team. Thanks for the info, I look forward to contributing more in the future

shanemcd commented:
@miles-w-3 Just pushed out 0.25.0 that includes your code. Thanks again!

nickjmv commented:
@miles-w-3, any idea when you might have time to create that ingress template you were talking about? ;)

nickjmv commented:
Nothing particular, just the basics to give it a custom DNS 😄

nickjmv commented:
@miles-w-3, not to push you, but let me know the PR you will create for that ingress, I like to see/follow the code to learn.

miles-w-3 commented:
Starting now, I will link the PR here
